File: exercises/Exercise049.java
public class Exercise049 {

    public static void main(String[] args) {
        boolean[] sieve = primeSieve(9999);
        int n2, n3;
        for (int n1 = 1000; n1 <= 3340; n1++)
        {
            n2 = n1 + 3330;
            n3 = n2 + 3330;
            if (sieve[n1 - 1] && sieve[n2 - 1] && sieve[n3 - 1])
            {
                if (arePermutations(n1, n2) && arePermutations(n2, n3))
                {
                    System.out.println(Integer.toString(n1) + Integer.toString(n2) + Integer.toString(n3));
                }
            }
        }
    }

    /**
     * This method determines whether two integers are permuations of each other.
     * @param n1 the first number.
     * @param n2 the second number.
     * @return true iff n1 is a permutation of n2.
     */
    public static boolean arePermutations(long n1, long n2)
    {
        int[] digits1 = new int[10];
        int[] digits2 = new int[10];
        for (char c1 : Long.toString(n1).toCharArray())
            digits1[c1 - '0']++;
        for (char c2 : Long.toString(n2).toCharArray())
            digits2[c2 - '0']++;
        for (int digit = 0; digit <= 9; digit++)
        {
            if (digits1[digit] != digits2[digit]) return false;
        }
        return true;
    }

    /**
     * This method returns the prime sieve of all numbers up to a certain limit;
     * @param limit the limit of the prime sieve.
     * @return a boolean[] representing the sieve up to limit.
     */
    public static boolean[] primeSieve(int limit)
    {
        if (limit <= 0) throw new IllegalArgumentException("Limit must be >= 1");

        // limit >= 1 here
        boolean[] sieve = new boolean[limit];
        for (int i = 2; i <= limit; i++) sieve[i-1] = true;

        for (int i = 2; i <= Math.sqrt(limit); i++)
        {
            if (sieve[i-1])
            {
                int j = i * i;
                while (j <= limit)
                {
                    sieve[j-1] = false;
                    j += i;
                }
            }
        }

        return sieve;
    }
}

File: python/exercise049.py
import math


# This method returns the prime sieve of all numbers up to a certain limit;
# limit the limit of the prime sieve.
# return a boolean[] representing the sieve up to limit.
def prime_sieve(limit):
    # limit >= 1 here
    sieve = [False]
    for i in range(2, limit + 1):
        sieve.append(True)

    for i in range(2, int(math.sqrt(limit)) + 1):
        if sieve[i-1]:
            j = i * i
            while j <= limit:
                sieve[j-1] = False
                j += i
    return sieve


# This method determines whether two integers are permuations of each other.
# n1 the first number.
# n2 the second number.
# return true iff n1 is a permutation of n2.
def are_permutations(n1, n2):
    digits1 = [0 for i in range(10)]
    digits2 = [0 for i in range(10)]
    for c in str(n1):
        digits1[int(c)] = digits1[int(c)] + 1
    for c in str(n2):
        digits2[int(c)] = digits2[int(c)] + 1
    for digit in range(10):
        if digits1[digit] != digits2[digit]:
            return False
    return True


sieve = prime_sieve(9_999)
for n1 in range(1000, 3340):
    n2 = n1 + 3330
    n3 = n2 + 3330
    if sieve[n1 - 1] is True and sieve[n2 - 1] is True and sieve[n3 - 1] is True:
        if are_permutations(n1, n2) and are_permutations(n2, n3):
            print(str(n1) + str(n2) + str(n3))